File: AudioKit/Operations/Signal Generators/Segment Arrays/AKSegmentArrayLoop.m
//
//  AKSegmentArrayLoop.m
//  AudioKit
//
//  Created by Aurelius Prochazka on 1/14/15
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//
//  Implementation of Csound's looptseg:
//  http://www.csounds.com/manual/html/looptseg.html
//

#import "AKSegmentArrayLoop.h"

@implementation AKSegmentArrayLoop
{
    AKParameter *loopFrequency;
    NSMutableArray<AKParameter *> *segments;
}

- (instancetype)initWithFrequency:(AKParameter *)frequency
                     initialValue:(AKParameter *)initialValue;
{
    self = [super initWithString:[self operationName]];
    if (self) {
        loopFrequency = frequency;
        segments = [[NSMutableArray alloc] init];
        [segments addObject:initialValue];
        self.state = @"connectable";
        self.dependencies = @[loopFrequency, initialValue];
    }
    return self;
}

- (void)addValue:(AKParameter *)value
   afterDuration:(AKParameter *)duration
       concavity:(AKParameter *)concavity
{
    [segments addObject:concavity];
    [segments addObject:duration];
    [segments addObject:value];
    self.dependencies = [self.dependencies arrayByAddingObjectsFromArray:@[concavity, duration, value]];
}

- (NSString *)stringForCSD {
    NSMutableArray *convertedSegments = [NSMutableArray array];
    for (AKParameter *parameter in segments) {
        [convertedSegments addObject:[NSString stringWithFormat:@"AKControl(%@)", parameter]];
    }
    return [NSString stringWithFormat:
            @"%@ looptseg %@, 0, 0, %@",
            self,
            loopFrequency,
            [convertedSegments componentsJoinedByString:@", "]];
}

@end

File: Tests/Tests/AKSegmentArrayLoop.swift
//
//  main.swift
//  AudioKit
//
//  Created by Aurelius Prochazka on 1/4/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

import Foundation

let testDuration: Float = 10;

class Instrument : AKInstrument {

    override init() {
        super.init()

        let segmentLoop = AKSegmentArrayLoop(
            frequency: 1.ak,
            initialValue: 440.ak
        )
        segmentLoop.addValue(550.ak, afterDuration: 1.ak, concavity: (-5).ak)
        segmentLoop.addValue(330.ak, afterDuration: 2.ak, concavity: 0.ak)
        segmentLoop.addValue(440.ak, afterDuration: 1.ak, concavity: 5.ak)

        enableParameterLog("segment.floatValue = ", parameter: segmentLoop, timeInterval: 0.1)

        let oscillator = AKOscillator()
        oscillator.frequency = segmentLoop
        setAudioOutput(oscillator)
    }
}

AKOrchestra.testForDuration(testDuration)

let instrument = Instrument()
AKOrchestra.addInstrument(instrument)

instrument.play()

NSThread.sleepForTimeInterval(NSTimeInterval(testDuration))
